### PR TITLE
First round of development

### PR DESCRIPTION
**Pull request by Dentosal**
Adds the following features:
* README.md with setup and usage instructions
* Protocol buffer definition
* gRPC communication
* mTLS certificate generation and use
* Simple job runner that can spawn jobs, stop them, and query status
* Job operation access limited only to the creating client
* CLI
* Some preliminary tests

Not in-scope yet:
* Resource limits and isolation
* Testing  more complex usage patterns and edge cases

### CODE REVIEW

File: client/src/bin/cli.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    match opts.subcmd {
+        Operation::Start { path, args } => {
+            let jobid = client.start(JobStartRequest { path, args }).await?;
+            println!("{}", jobid);
+        }
+        Operation::Stop { jobid } => client.stop(JobId::parse(&jobid)?).await?,
+        Operation::Status { jobid } => {
+            pretty_print_status(client.status(JobId::parse(&jobid)?).await?);
+        }
+        Operation::Wait { jobid } => {
```
Comment (knisbet) [inline client/src/bin/cli.rs:66]: I don't believe Wait was a listed endpoint in the design (was maybe a potential improvement at the end). Why the deviation from the design?

File: gen-certs.sh [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    -key root-ca.key -sha256 -days 1024 -out root-ca.crt
+openssl x509 -outform der -in root-ca.crt -out root-ca.der
+cd ..
+
+
+# Generate server certificate
+
+cd server
+openssl genrsa -out server.key 4096
+# openssl req -new -key server.key -subj '/CN=localhost/O=Server/C=US' -out server.csr
```
Comment (knisbet) [inline gen-certs.sh]: nitpick: if this isn't needed should probably be cleaned up before merge.

File: server/src/output_stream.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+/// Start a task that streams from an `OutputHandler` to a mpsc channel.
+pub fn stream_to(from: Arc<OutputHandler>, to: Sender<Result<OutputEvent, tonic::Status>>) {
+    use std::borrow::Borrow;
+    tokio::spawn(async move {
+        let mut index = 0;
+        loop {
+            let h: &OutputHandler = from.borrow();
+            let state = h.state.read().await;
```
Comment (knisbet) [inline server/src/output_stream.rs]: I might be missing something since I don't have as deep rust experience as the other reviewers, but won't this hold the read lock through the .await calls below?

File: server/src/job.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+/// A single running job, i.e. a process
+pub struct Job {
+    pub owner: ClientName,
+    child: Child,
+    pub stdout: Arc<OutputHandler>,
+    pub stderr: Arc<OutputHandler>,
+}
+impl Job {
+    pub fn spawn(owner: ClientName, req: JobStartRequest) -> Result<Self, String> {
```
Comment (knisbet) [inline server/src/job.rs:38]: I'm not sure if I just can't find it, but I don't see anything but the status function trying to wait on the child. I suspect this will produce zombies if the client never checks the status of the child process via the API after the child exits.

File: server/src/output_stream.rs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                inner.push(x.to_vec()).await;
+            }
+            inner.complete().await;
+        });
+
+        arc_self
+    }
+
+    /// Push new data to the history, notifying all waiting processes
+    pub async fn push(&self, data: Vec<u8>) {
```
Comment (knisbet) [inline server/src/output_stream.rs:60]: Not sure push/complete should be public functions as they look internal only, especially since the process can panic if miss-used.

File: client/src/bin/cli.rs (hunk 1 above)
Comment (Dentosal) [reply client/src/bin/cli.rs:66]: You're right, it wasn't. When implementing tests, I noticed that it would be way more convenient if the CLI had a status command. By implementing it in the client library, it could be changed to a proper notification instead of loinng without chaning the API at all. However, if you think I should stick to the original design exactly, I can do that as well.

File: server/src/output_stream.rs (hunk 3 above)
Comment (Dentosal) [reply server/src/output_stream.rs]: Yes, and it will actually block in case of multiple readers. Fixed in 8c5f264. That's sadly a bit complicated to test automatically, but I maybe I should try.

File: server/src/output_stream.rs (hunk 5 above)
Comment (Dentosal) [reply server/src/output_stream.rs:60]: They are not necessarily internal only; if you would like to use `OutputHandler` for something that doesn't implement `AsyncRead` (for instance, stdlib process output), then those would be still useful. The panic conditions are documented in the doc comment. Although they are used here only internally, there is nothing from with them being public (from OOP perspective, anyway). On the other hand, it might be a bit cleaner to keep them private until they are used outside the class, since the whole `server` is a binary crate, and doesn't provide an code-level API anyway.

File: server/src/job.rs (hunk 4 above)
Comment (Dentosal) [reply server/src/job.rs:38]: You're right, this will cause zombie processes. I though that Tokio should be able to handle this by itself, as the documentation says

> The tokio runtime will, on a best-effort basis, attempt to reap and clean up any process which it has spawned. No additional guarantees are made with regards how quickly or how often this procedure will take place.

Looking more into it, this only concerns cases when the  `Child` handle has been dropped. Since this is not the case, the best-effort orphan reaper will not run.

I've fixed this in by 12782b1 managing the child process in a separate task.

File: gen-certs.sh (hunk 2 above)
Comment (Dentosal) [reply gen-certs.sh]: Cleanup in 254da17

File: server/src/output_stream.rs (hunk 3 above)
Comment (Dentosal) [reply server/src/output_stream.rs]: Added a test case for it, in 3236e05 under `common/tests/concurrent.rs`

File: client/src/bin/cli.rs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            client.output(JobId::parse(&jobid)?, tx).await?;
+
+            // If writing to stdout or stderr fails, there is no reason to continue => panic
+
+            let mut out = std::io::stdout();
+            let mut err = std::io::stderr();
+
+            while let Some(event) = rx.recv().await {
+                match event? {
+                    (OutputStream::Stdout, data) => out.write_all(&data).unwrap(),
```
Comment (xacrimon) [inline client/src/bin/cli.rs]: This can fail spontaneously due to OS stuff iirc. Should at least have proper error handling and reporting. Unwraps are only to be used when you can prove with a pen and piece of paper and some logic why something cannot panic.

File: client/src/bin/cli.rs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+            out.flush().unwrap();
+            err.flush().unwrap();
+        }
+    }
+
+    Ok(())
+}
+
+fn pretty_print_status(status: JobStatus) {
```
Comment (xacrimon) [inline client/src/bin/cli.rs]: To follow Rust idioms and guidelines this should be in the form of an implementation of `std::fmt::Display` on `JobStatus`.

File: client/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+// TODO: create a proper error type, and add conversions to it
+
+type DResult<T> = Result<T, Box<dyn std::error::Error>>;
+
+pub struct Client {
+    client: TServiceClient<Channel>,
+}
+impl Client {
+    /// Connect to a job server
+    pub async fn connect(target: String, tls: TlsConfig) -> DResult<Self> {
```
Comment (xacrimon) [inline client/src/lib.rs]: This should take &str if possible.

File: client/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    /// Starts a new job by spawning a process from given executable path and arguments.
+    /// Returns a unique job id (UUID v4), that is used to specify the target job for other endpoints.
+    // If the executable is not found or cannot be executed, immediately returns an error.
+    pub async fn start(&mut self, req: JobStartRequest) -> DResult<JobId> {
+        let command = tonic::Request::new(req);
+        let response = self.client.start(command).await?;
+        Ok(response
+            .into_inner()
+            .try_into()
+            .expect("Server returned invalid JobId"))
```
Comment (xacrimon) [inline client/src/lib.rs]: Lacks proper error handling.

File: client/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    /// Cancels a job by sending `SIGKILL`.
+    /// This is done asynchronously, and can return before the process has terminated.
+    /// If you must wait until the job has stopped, it can do so by calling `wait`.
```
Comment (xacrimon) [inline client/src/lib.rs]: Calling sigkill on a stopped job makes no sense.

File: client/src/bin/cli.rs (hunk 6 above)
Comment (Dentosal) [reply client/src/bin/cli.rs]: > This can fail spontaneously due to OS stuff iirc.

Yes, it can fail. However, it isn't much use to report errors after that. Where would the report go? To the stdout we cannot write to? I agree that `expect` would still be more readable, and will change to that.

Comment (Dentosal) [reply client/src/bin/cli.rs]: Fixed in 919732f

File: client/src/bin/cli.rs (hunk 7 above)
Comment (Dentosal) [reply client/src/bin/cli.rs]: Fixed in 93bd033.

File: client/src/lib.rs (hunk 9 above)
Comment (Dentosal) [reply client/src/lib.rs]: This is an unrecoverable critical error. The panic should be acceptable here, unless we anticipate that the communication format will be changing in the future. But then again, we do not have any other version negotiation or forward-compatibility either.

File: client/src/lib.rs (hunk 8 above)
Comment (Dentosal) [reply client/src/lib.rs]: The internal client requires owned `String` anyway, so I'm going to use `Into<String>` to accept both. Fixed in 6304652.

File: client/src/lib.rs (hunk 10 above)
Comment (Dentosal) [reply client/src/lib.rs]: Clarified in d5b1e53.

File: client/src/bin/cli.rs (hunk 1 above)
Comment (knisbet) [reply client/src/bin/cli.rs:66]: We're not sticklers to the design at Teleport, there will always be new ideas or things that come up during implementation, I just wanted to understand why the change was introduced.

File: client/src/lib.rs (hunk 9 above)
Comment (Dentosal) [reply client/src/lib.rs]: Fixed in 83f768f: Added a better error message, which is returned to the client.

**Review by xacrimon — APPROVED** [review]

File: server/src/output_stream.rs [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+                    break;
+                }
+
+                index += 1;
+            } else if completed {
+                // All content has been streamed
+                break;
+            } else {
+                // Wait until more output is available
+                h.notify.notified().await;
```
Comment (knisbet) [inline server/src/output_stream.rs]: What happens if data is pushed to the history between dropping the read lock and registering the waiting task in notify... and then there isn't new data / a new notification for awhile?

Comment (Dentosal) [reply server/src/output_stream.rs]: That causes a delay until more output is pushed. In worst case the client registers after the job has completed, causing the client to wait indefinitely.

This is clearly a bug. Fixed in f44a564.

Unfortunately, [`tokio::sync`](https://docs.rs/tokio/1.7.0/tokio/sync/index.html) seems to offer no ready solutions to this. Instread, I rolled my own synchronization method, internally using an atomic generation number, and a `oneshot`-channel based update-notification registration.

File: server/src/output_stream/handler.rs [hunk 12]
```diff
@@ -1,35 +1,54 @@
 use std::sync::Arc;
 
 use tokio::sync::mpsc::Sender;
-use tokio::sync::{Notify, RwLock};
```
Comment (knisbet) [inline server/src/output_stream/handler.rs]: I think notify would've worked for you, you just need to split up when you call notified() and when you call await, as per https://docs.rs/tokio/1.7.0/tokio/sync/struct.Notify.html#method.notify_waiters

Comment (Dentosal) [reply server/src/output_stream/handler.rs]: Right, reading the documentation. Should have done that. 

New attempt here, this time only moving the creation around 52c2043
